File: priority_engine.js
/*
 * priority_engine.js — APH State Funding Priority Model Scoring Engine (JS Port)
 * ================================================================================
 *
 * Browser-side scoring engine. API-equivalent to priority_engine.py so the
 * configurator UI can update rankings live as the user changes weights without
 * a server round-trip.
 *
 * Engine equivalence test (sprint 6) compares Python and JS outputs against
 * the same configs and projects.
 *
 * Reference: ACRP Synthesis 123. Reference state encodings in schema/reference_configs/.
 *
 * © 2026 John Marcus Cocanougher. All rights reserved.
 *
 * Exposes:
 *   PriorityModel — load + score
 *   loadProjectsFromCSV — CSV string → array of objects
 *   safeEval — arithmetic expression evaluator (no JS eval)
 *
 * Both UMD (browser globals) and ES module exports are supported.
 */

(function (root, factory) {
  if (typeof exports === "object" && typeof module === "object") {
    module.exports = factory();
  } else if (typeof define === "function" && define.amd) {
    define([], factory);
  } else {
    root.PriorityEngine = factory();
  }
})(typeof self !== "undefined" ? self : this, function () {
  "use strict";

  const SCHEMA_VERSION = "1.0.0";

  // ── Safe expression evaluator ─────────────────────────────────────
  // Recursive-descent parser for arithmetic with named variables.
  // Supports + - * / // % ** parentheses unary -/+

  function tokenize(expr) {
    const tokens = [];
    let i = 0;
    while (i < expr.length) {
      const c = expr[i];
      if (/\s/.test(c)) { i++; continue; }
      if (/[0-9.]/.test(c)) {
        let j = i;
        while (j < expr.length && /[0-9.]/.test(expr[j])) j++;
        tokens.push({ type: "num", value: parseFloat(expr.slice(i, j)) });
        i = j;
      } else if (/[A-Za-z_]/.test(c)) {
        let j = i;
        while (j < expr.length && /[A-Za-z0-9_]/.test(expr[j])) j++;
        tokens.push({ type: "ident", value: expr.slice(i, j) });
        i = j;
      } else if ("+-*/%()".includes(c)) {
        // Detect ** and //
        if ((c === "*" && expr[i + 1] === "*") || (c === "/" && expr[i + 1] === "/")) {
          tokens.push({ type: "op", value: c + expr[i + 1] });
          i += 2;
        } else {
          tokens.push({ type: "op", value: c });
          i++;
        }
      } else {
        throw new Error("Unexpected character in formula: " + c);
      }
    }
    return tokens;
  }

  function safeEval(expr, vars) {
    const tokens = tokenize(expr);
    let pos = 0;
    function peek() { return tokens[pos]; }
    function eat() { return tokens[pos++]; }

    function parseExpr() {
      let left = parseTerm();
      while (peek() && peek().type === "op" && (peek().value === "+" || peek().value === "-")) {
        const op = eat().value;
        const right = parseTerm();
        left = op === "+" ? left + right : left - right;
      }
      return left;
    }

    function parseTerm() {
      let left = parseFactor();
      while (peek() && peek().type === "op" &&
             (peek().value === "*" || peek().value === "/" ||
              peek().value === "//" || peek().value === "%")) {
        const op = eat().value;
        const right = parseFactor();
        if (op === "*") left = left * right;
        else if (op === "/") left = left / right;
        else if (op === "//") left = Math.floor(left / right);
        else if (op === "%") left = left % right;
      }
      return left;
    }

    function parseFactor() {
      let base = parseUnary();
      if (peek() && peek().type === "op" && peek().value === "**") {
        eat();
        const exp = parseFactor();
        return Math.pow(base, exp);
      }
      return base;
    }

    function parseUnary() {
      if (peek() && peek().type === "op" && (peek().value === "-" || peek().value === "+")) {
        const op = eat().value;
        const v = parseUnary();
        return op === "-" ? -v : v;
      }
      return parsePrimary();
    }

    function parsePrimary() {
      const t = eat();
      if (!t) throw new Error("Unexpected end of formula");
      if (t.type === "num") return t.value;
      if (t.type === "ident") {
        if (!(t.value in vars)) throw new Error("Unknown variable: " + t.value);
        return vars[t.value];
      }
      if (t.type === "op" && t.value === "(") {
        const v = parseExpr();
        const close = eat();
        if (!close || close.value !== ")") throw new Error("Missing closing paren");
        return v;
      }
      throw new Error("Unexpected token: " + JSON.stringify(t));
    }

    const result = parseExpr();
    if (pos !== tokens.length) throw new Error("Trailing input in formula");
    return result;
  }

  // ── Coercion ─────────────────────────────────────────────────────
  function coerce(v) {
    if (typeof v !== "string") return v;
    const s = v.trim();
    if (s.toLowerCase() === "true" || s.toLowerCase() === "yes") return true;
    if (s.toLowerCase() === "false" || s.toLowerCase() === "no") return false;
    if (/^-?\d+$/.test(s)) return parseInt(s, 10);
    if (/^-?\d+\.\d+$/.test(s)) return parseFloat(s);
    return s;
  }

  // ── Field-match operators ───────────────────────────────────────
  const COMPARATORS = {
    eq:     (a, b) => a == b,
    ne:     (a, b) => a != b,
    in:     (a, b) => Array.isArray(b) && b.includes(a),
    not_in: (a, b) => Array.isArray(b) && !b.includes(a),
    gte:    (a, b) => Number(a) >= Number(b),
    lte:    (a, b) => Number(a) <= Number(b),
    gt:     (a, b) => Number(a) > Number(b),
    lt:     (a, b) => Number(a) < Number(b),
    matches: (a, b) => new RegExp(b).test(String(a)),
  };

  function evaluateFieldMatch(fm, project) {
    const cmp = COMPARATORS[fm.operator];
    if (!cmp) throw new Error("Unknown operator: " + fm.operator);
    const actual = coerce(project[fm.field]);
    try { return cmp(actual, fm.value); } catch { return false; }
  }

  // ── Main engine class ───────────────────────────────────────────
  class PriorityModel {
    constructor(config) {
      if (config.schema_version !== SCHEMA_VERSION) {
        throw new Error(
          `Schema version mismatch: expected ${SCHEMA_VERSION}, got ${config.schema_version}`
        );
      }
      this.config = config;
      this.metadata = config.metadata;
      this.weightingModel = config.weighting_model;
      this.criteria = config.criteria;
      this.eligibilityGate = config.eligibility_gate || { enabled: false };
      this.setAsides = config.set_asides || [];
      this.adjustments = config.adjustments || {};
      this.governance = config.governance;
    }

    static fromConfig(config) { return new PriorityModel(config); }

    // ── Eligibility gate ──────────────────────────────────────────
    checkEligibility(project) {
      if (!this.eligibilityGate.enabled) return { passed: true, failures: [] };
      const failures = [];
      for (const f of this.eligibilityGate.filters || []) {
        if (f.field_check && !evaluateFieldMatch(f.field_check, project)) {
          if ((f.on_fail || "exclude") === "exclude") {
            failures.push(`${f.label}: ${f.rule}`);
          }
        }
      }
      return { passed: failures.length === 0, failures };
    }

    subcategoryMatches(criterion, sub, project) {
      if (sub.field_match) return evaluateFieldMatch(sub.field_match, project);
      const fieldName = criterion.project_field || criterion.id;
      const projVal = project[fieldName];
      if (projVal === null || projVal === undefined || projVal === "") return false;
      return String(projVal).trim() === sub.id;
    }

    scoreCriterion(criterion, project) {
      const cid = criterion.id;
      const clabel = criterion.label;
      const bucket = criterion.bucket || null;
      const style = criterion.scoring_style;
      const maxPts = criterion.max_points;
      const weight = criterion.weight !== undefined ? criterion.weight : 1;
      const subs = criterion.subcategories || [];

      let matchedIds = [];
      let rawScore = 0;
      let explanation = "";

      if (style === "formula") {
        const expr = criterion.formula || "0";
        const varsMap = {};
        subs.forEach(s => {
          if (this.subcategoryMatches(criterion, s, project)) varsMap[s.id] = s.score;
        });
        try {
          rawScore = safeEval(expr, varsMap);
          explanation = `formula '${expr}' = ${rawScore}`;
        } catch (e) {
          rawScore = 0;
          explanation = `formula error: ${e.message}`;
        }
      } else if (style === "select_one") {
        const candidates = subs.filter(s => this.subcategoryMatches(criterion, s, project));
        if (candidates.length) {
          const best = candidates.reduce((a, b) => a.score >= b.score ? a : b);
          matchedIds = [best.id];
          rawScore = best.score;
          explanation = `matched ${best.id} (score ${best.score})`;
        } else {
          explanation = "no subcategory matched";
        }
      } else if (style === "tiered_score") {
        const hit = subs.find(s => this.subcategoryMatches(criterion, s, project));
        if (hit) {
          matchedIds = [hit.id];
          rawScore = hit.score;
          explanation = `matched tier ${hit.id} (score ${hit.score})`;
        } else {
          explanation = "no tier matched";
        }
      } else if (style === "select_many") {
        const matched = subs.filter(s => this.subcategoryMatches(criterion, s, project));
        matchedIds = matched.map(s => s.id);
        rawScore = matched.reduce((sum, s) => sum + s.score, 0);
        explanation = matched.length
          ? `summed ${matched.length} matches: ${matchedIds.join(", ")}`
          : "no subcategories matched";
      } else if (style === "accumulating") {
        const matched = subs.filter(s => this.subcategoryMatches(criterion, s, project));
        matchedIds = matched.map(s => s.id);
        rawScore = matched.reduce((sum, s) => sum + s.score, 0);
        if (maxPts !== undefined && rawScore > maxPts) rawScore = maxPts;
        explanation = matched.length
          ? `accumulated ${matched.length} matches (cap ${maxPts}): ${matchedIds.join(", ")}`
          : "no subcategories matched";
      } else if (style === "flat_lookup") {
        const hit = subs.find(s => this.subcategoryMatches(criterion, s, project));
        if (hit) {
          matchedIds = [hit.id];
          rawScore = hit.score;
          explanation = `lookup hit ${hit.id} (score ${hit.score})`;
        } else {
          explanation = "no lookup match for project_type";
        }
      } else {
        throw new Error("Unknown scoring_style: " + style);
      }

      let weighted = rawScore;
      if (this.weightingModel === "multiplicative") {
        weighted = rawScore * weight;
      } else if (this.weightingModel === "additive_100pt" && maxPts !== undefined) {
        if (weighted < 0) weighted = Math.max(weighted, -Math.abs(maxPts));
        else weighted = Math.min(weighted, maxPts);
      }

      return {
        criterion_id: cid,
        criterion_label: clabel,
        bucket,
        matched_subcategory_ids: matchedIds,
        raw_score: rawScore,
        weighted_score: weighted,
        explanation,
      };
    }

    applyAdjustments(base, project) {
      const applied = {};
      const notes = [];
      let score = base;

      const rev = this.adjustments.revenue_producing_bump || {};
      if (rev.enabled) {
        const tags = rev.applies_to_project_types || [];
        const ptype = String(project.project_type || "");
        if (tags.some(t => ptype.includes(t))) {
          const bonus = rev.bonus_amount || 0;
          score += bonus;
          applied.revenue_producing_bump = bonus;
          notes.push(`revenue_producing_bump +${bonus}`);
        }
      }

      const lfp = this.adjustments.low_federal_priority_bonus || {};
      if (lfp.enabled) {
        const nprs = coerce(project.nprs_score);
        if (typeof nprs === "number" && nprs <= (lfp.nprs_threshold || 0)) {
          const bonus = lfp.bonus_amount || 0;
          score += bonus;
          applied.low_federal_priority_bonus = bonus;
          notes.push(`low_federal_priority_bonus +${bonus}`);
        }
      }

      return { score, applied, notes };
    }

    assignSetAside(project) {
      for (const sa of this.setAsides) {
        if (sa.field_filter && evaluateFieldMatch(sa.field_filter, project)) return sa.id;
      }
      return null;
    }

    scoreProject(project) {
      const audit = [];
      const elig = this.checkEligibility(project);
      if (!elig.passed) audit.push("FAILED eligibility gate: " + elig.failures.join("; "));

      const contributions = this.criteria.map(c => {
        const contrib = this.scoreCriterion(c, project);
        audit.push(
          `  ${contrib.criterion_id}: raw=${contrib.raw_score} ` +
          `weighted=${contrib.weighted_score} (${contrib.explanation})`
        );
        return contrib;
      });

      let baseScore;
      if (this.weightingModel === "formula") {
        const expr = this.config.global_formula || "0";
        const vars = {};
        contributions.forEach(c => { vars[c.criterion_id] = c.raw_score; });
        try { baseScore = safeEval(expr, vars); audit.push(`global formula '${expr}' = ${baseScore}`); }
        catch (e) { baseScore = 0; audit.push(`global formula error: ${e.message}`); }
      } else {
        baseScore = contributions.reduce((s, c) => s + c.weighted_score, 0);
      }

      const adj = this.applyAdjustments(baseScore, project);
      audit.push(...adj.notes);

      const sa = elig.passed ? this.assignSetAside(project) : null;
      if (sa) audit.push(`set-aside assignment: ${sa}`);

      return {
        project_id: String(project.project_id ?? "?"),
        project_label: String(project.description || project.airport_name || project.project_id || "?"),
        eligibility_passed: elig.passed,
        eligibility_failures: elig.failures,
        base_score: baseScore,
        contributions,
        adjustments_applied: adj.applied,
        set_aside_assignment: sa,
        final_score: elig.passed ? adj.score : 0,
        audit_trail: audit,
      };
    }

    rank(projects) {
      const scored = [], excluded = [];
      for (const p of projects) {
        const s = this.scoreProject(p);
        (s.eligibility_passed ? scored : excluded).push(s);
      }
      return {
        model_name: this.metadata.name,
        model_state: this.metadata.state,
        weighting_model: this.weightingModel,
        scored,
        excluded,
        ranked: (descending = true) =>
          [...scored].sort((a, b) => descending ? b.final_score - a.final_score : a.final_score - b.final_score),
      };
    }
  }

  // ── CSV loader (minimal) ────────────────────────────────────────
  function loadProjectsFromCSV(csvText) {
    const lines = csvText.split(/\r?\n/).filter(l => l.length);
    if (!lines.length) return [];
    const headers = lines[0].split(",").map(h => h.trim());
    const rows = [];
    for (let i = 1; i < lines.length; i++) {
      const cells = lines[i].split(",");
      const obj = {};
      headers.forEach((h, idx) => obj[h] = coerce((cells[idx] || "").trim()));
      rows.push(obj);
    }
    return rows;
  }

  return {
    PriorityModel,
    loadProjectsFromCSV,
    safeEval,
    SCHEMA_VERSION,
  };
});

File: priority_engine.py
"""
priority_engine.py — APH State Funding Priority Model Scoring Engine
=====================================================================

Pure-Python scoring engine for state aviation funding priority models.
No APH dependencies. Importable as a library, runnable via priority_cli.py.

Architectural sibling of tools/aip_eligibility_tool/aip_engine.py — that one
implements federal NPRS scoring; this one implements configurable state-level
prioritization grounded in ACRP Synthesis 123.

Reference: ACRP Synthesis 123 — State Aviation Funding: Project Prioritization
and Selection Processes (2023). Reference state encodings: Wyoming PRM 2021,
MnDOT, ALDOT, Louisiana DOTD.

© 2026 John Marcus Cocanougher. All rights reserved.
"""

from __future__ import annotations

import ast
import csv
import json
import operator
from dataclasses import dataclass, field, asdict
from pathlib import Path
from typing import Any, Optional, Union

# ═══════════════════════════════════════════════════════════════════
#  TYPES
# ═══════════════════════════════════════════════════════════════════

JSON = dict[str, Any]
Project = dict[str, Any]


@dataclass
class CriterionContribution:
    """How one criterion contributed to a project's total score."""
    criterion_id: str
    criterion_label: str
    bucket: Optional[str]
    matched_subcategory_ids: list[str]
    raw_score: float
    weighted_score: float
    explanation: str


@dataclass
class ProjectScore:
    """Full scoring output for one project, including audit trail."""
    project_id: str
    project_label: str
    eligibility_passed: bool
    eligibility_failures: list[str]
    base_score: float
    contributions: list[CriterionContribution]
    adjustments_applied: dict[str, float]
    set_aside_assignment: Optional[str]
    final_score: float
    audit_trail: list[str]

    def to_dict(self) -> JSON:
        d = asdict(self)
        return d


@dataclass
class RankedResult:
    """Output of a model run against a project list."""
    model_name: str
    model_state: str
    weighting_model: str
    scored: list[ProjectScore]
    excluded: list[ProjectScore]  # failed eligibility gate

    def ranked(self, descending: bool = True) -> list[ProjectScore]:
        return sorted(self.scored, key=lambda p: p.final_score, reverse=descending)


# ═══════════════════════════════════════════════════════════════════
#  SAFE EXPRESSION EVALUATOR (for formula scoring)
# ═══════════════════════════════════════════════════════════════════

_SAFE_OPS: dict[type, Any] = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Div: operator.truediv,
    ast.FloorDiv: operator.floordiv,
    ast.Mod: operator.mod,
    ast.Pow: operator.pow,
    ast.USub: operator.neg,
    ast.UAdd: operator.pos,
}


def safe_eval(expr: str, variables: dict[str, float]) -> float:
    """Evaluate an arithmetic expression with named variables. Never uses raw eval()."""
    tree = ast.parse(expr, mode="eval")

    def _eval(node):
        if isinstance(node, ast.Expression):
            return _eval(node.body)
        if isinstance(node, ast.Constant) and isinstance(node.value, (int, float)):
            return node.value
        if isinstance(node, ast.Name):
            if node.id not in variables:
                raise ValueError(f"Unknown variable in formula: {node.id}")
            return variables[node.id]
        if isinstance(node, ast.BinOp):
            op = _SAFE_OPS.get(type(node.op))
            if op is None:
                raise ValueError(f"Operator not allowed: {type(node.op).__name__}")
            return op(_eval(node.left), _eval(node.right))
        if isinstance(node, ast.UnaryOp):
            op = _SAFE_OPS.get(type(node.op))
            if op is None:
                raise ValueError(f"Unary op not allowed: {type(node.op).__name__}")
            return op(_eval(node.operand))
        raise ValueError(f"Disallowed expression node: {type(node).__name__}")

    return float(_eval(tree))


# ═══════════════════════════════════════════════════════════════════
#  FIELD-MATCH EVALUATION
# ═══════════════════════════════════════════════════════════════════

_COMPARATORS = {
    "eq": lambda a, b: a == b,
    "ne": lambda a, b: a != b,
    "in": lambda a, b: a in b,
    "not_in": lambda a, b: a not in b,
    "gte": lambda a, b: float(a) >= float(b),
    "lte": lambda a, b: float(a) <= float(b),
    "gt": lambda a, b: float(a) > float(b),
    "lt": lambda a, b: float(a) < float(b),
    "matches": lambda a, b: __import__("re").search(b, str(a)) is not None,
}


def _coerce(value: Any) -> Any:
    """Coerce CSV-string values to bool/number where appropriate."""
    if not isinstance(value, str):
        return value
    s = value.strip()
    if s.lower() in ("true", "yes"):
        return True
    if s.lower() in ("false", "no"):
        return False
    try:
        if "." in s:
            return float(s)
        return int(s)
    except ValueError:
        return s


def evaluate_field_match(field_match: JSON, project: Project) -> bool:
    fld = field_match["field"]
    op = field_match["operator"]
    expected = field_match["value"]
    actual = _coerce(project.get(fld))
    cmp = _COMPARATORS.get(op)
    if cmp is None:
        raise ValueError(f"Unknown operator: {op}")
    try:
        return cmp(actual, expected)
    except (TypeError, ValueError):
        return False


# ═══════════════════════════════════════════════════════════════════
#  ENGINE
# ═══════════════════════════════════════════════════════════════════

class PriorityModel:
    """A loaded priority model, ready to score projects."""

    SCHEMA_VERSION = "1.0.0"

    def __init__(self, config: JSON):
        if config.get("schema_version") != self.SCHEMA_VERSION:
            raise ValueError(
                f"Schema version mismatch: expected {self.SCHEMA_VERSION}, "
                f"got {config.get('schema_version')}"
            )
        self.config = config
        self.metadata = config["metadata"]
        self.weighting_model = config["weighting_model"]
        self.criteria = config["criteria"]
        self.eligibility_gate = config.get("eligibility_gate", {"enabled": False})
        self.set_asides = config.get("set_asides", [])
        self.adjustments = config.get("adjustments", {})
        self.governance = config["governance"]

    @classmethod
    def from_file(cls, path: Union[str, Path]) -> "PriorityModel":
        return cls(json.loads(Path(path).read_text()))

    # ── Eligibility gate ─────────────────────────────────────────────
    def check_eligibility(self, project: Project) -> tuple[bool, list[str]]:
        if not self.eligibility_gate.get("enabled"):
            return True, []
        failures = []
        for f in self.eligibility_gate.get("filters", []):
            fc = f.get("field_check")
            if fc and not evaluate_field_match(fc, project):
                if f.get("on_fail", "exclude") == "exclude":
                    failures.append(f"{f['label']}: {f['rule']}")
        return (len(failures) == 0), failures

    # ── Subcategory matching ─────────────────────────────────────────
    def _subcategory_matches(self, criterion: JSON, sub: JSON, project: Project) -> bool:
        """Does this subcategory match this project under this criterion?"""
        # 1. Explicit field_match on the subcategory wins
        if "field_match" in sub:
            return evaluate_field_match(sub["field_match"], project)

        # 2. Otherwise look up the criterion's project_field (or fall back to criterion id)
        field_name = criterion.get("project_field") or criterion["id"]
        proj_value = project.get(field_name)
        if proj_value is None or proj_value == "":
            return False
        return str(proj_value).strip() == sub["id"]

    # ── Per-criterion scoring ────────────────────────────────────────
    def _score_criterion(
        self, criterion: JSON, project: Project
    ) -> CriterionContribution:
        cid = criterion["id"]
        clabel = criterion["label"]
        bucket = criterion.get("bucket")
        style = criterion["scoring_style"]
        max_points = criterion.get("max_points")
        weight = criterion.get("weight", 1)

        matched_ids: list[str] = []
        raw_score: float = 0.0
        explanation: str

        subs = criterion.get("subcategories", [])

        if style == "formula":
            expr = criterion.get("formula", "0")
            # Variables: each subcategory id maps to its score if matched, else 0
            vars_map = {
                s["id"]: float(s["score"])
                for s in subs
                if self._subcategory_matches(criterion, s, project)
            }
            try:
                raw_score = safe_eval(expr, vars_map)
                explanation = f"formula '{expr}' = {raw_score}"
            except Exception as e:
                raw_score = 0
                explanation = f"formula evaluation error: {e}"

        elif style == "select_one":
            # If multiple subcategories match (e.g., gte tiers), take highest score
            candidates = [s for s in subs if self._subcategory_matches(criterion, s, project)]
            if candidates:
                best = max(candidates, key=lambda s: s["score"])
                matched_ids = [best["id"]]
                raw_score = float(best["score"])
                explanation = f"matched {best['id']} (score {best['score']})"
            else:
                explanation = "no subcategory matched"

        elif style == "tiered_score":
            # Mutually exclusive — first match wins
            for sub in subs:
                if self._subcategory_matches(criterion, sub, project):
                    matched_ids = [sub["id"]]
                    raw_score = float(sub["score"])
                    explanation = f"matched tier {sub['id']} (score {sub['score']})"
                    break
            else:
                explanation = "no tier matched"

        elif style == "select_many":
            # Sum all matching subcategory scores
            matched = [s for s in subs if self._subcategory_matches(criterion, s, project)]
            matched_ids = [s["id"] for s in matched]
            raw_score = sum(float(s["score"]) for s in matched)
            explanation = (
                f"summed {len(matched)} matches: " + ", ".join(matched_ids)
                if matched else "no subcategories matched"
            )

        elif style == "accumulating":
            # Like select_many but capped at max_points
            matched = [s for s in subs if self._subcategory_matches(criterion, s, project)]
            matched_ids = [s["id"] for s in matched]
            raw_score = sum(float(s["score"]) for s in matched)
            if max_points is not None and raw_score > max_points:
                raw_score = max_points
            explanation = (
                f"accumulated {len(matched)} matches (capped at {max_points}): "
                + ", ".join(matched_ids)
                if matched else "no subcategories matched"
            )

        elif style == "flat_lookup":
            # One subcategory match — uses the criterion's project_field for direct lookup
            for sub in subs:
                if self._subcategory_matches(criterion, sub, project):
                    matched_ids = [sub["id"]]
                    raw_score = float(sub["score"])
                    explanation = f"lookup hit {sub['id']} (score {sub['score']})"
                    break
            else:
                explanation = f"no lookup match for project_type"

        else:
            raise ValueError(f"Unknown scoring_style: {style}")

        # Apply weighting based on the model's overall weighting_model
        if self.weighting_model == "multiplicative":
            weighted = raw_score * float(weight)
        elif self.weighting_model == "additive_100pt":
            weighted = raw_score
            if max_points is not None:
                # Cap contribution at max_points (allows negatives below 0)
                weighted = max(weighted, -abs(max_points)) if weighted < 0 else min(weighted, max_points)
        elif self.weighting_model == "formula":
            # Per-criterion weighting handled inside global formula at caller level
            weighted = raw_score
        elif self.weighting_model == "dual_objective_subjective":
            weighted = raw_score
        else:
            weighted = raw_score

        return CriterionContribution(
            criterion_id=cid,
            criterion_label=clabel,
            bucket=bucket,
            matched_subcategory_ids=matched_ids,
            raw_score=raw_score,
            weighted_score=weighted,
            explanation=explanation,
        )

    # ── Apply post-scoring adjustments ───────────────────────────────
    def _apply_adjustments(
        self, base: float, project: Project
    ) -> tuple[float, dict[str, float], list[str]]:
        adj_applied: dict[str, float] = {}
        notes: list[str] = []
        score = base

        rev = self.adjustments.get("revenue_producing_bump", {}) or {}
        if rev.get("enabled"):
            applies_to = rev.get("applies_to_project_types", [])
            ptype = project.get("project_type")
            if any(tag in str(ptype) for tag in applies_to):
                bonus = float(rev.get("bonus_amount", 0))
                score += bonus
                adj_applied["revenue_producing_bump"] = bonus
                notes.append(f"revenue_producing_bump +{bonus}")

        lfp = self.adjustments.get("low_federal_priority_bonus", {}) or {}
        if lfp.get("enabled"):
            nprs = _coerce(project.get("nprs_score"))
            if isinstance(nprs, (int, float)):
                if nprs <= float(lfp.get("nprs_threshold", 0)):
                    bonus = float(lfp.get("bonus_amount", 0))
                    score += bonus
                    adj_applied["low_federal_priority_bonus"] = bonus
                    notes.append(f"low_federal_priority_bonus +{bonus}")

        return score, adj_applied, notes

    # ── Set-aside assignment ─────────────────────────────────────────
    def _assign_set_aside(self, project: Project) -> Optional[str]:
        for sa in self.set_asides:
            ff = sa.get("field_filter")
            if ff and evaluate_field_match(ff, project):
                return sa["id"]
        return None

    # ── Score one project ────────────────────────────────────────────
    def score_project(self, project: Project) -> ProjectScore:
        project_id = str(project.get("project_id", "?"))
        project_label = project.get("description") or project.get("airport_name") or project_id
        audit: list[str] = []

        # 1. Eligibility gate
        passed, failures = self.check_eligibility(project)
        if not passed:
            audit.append(f"FAILED eligibility gate: {'; '.join(failures)}")

        # 2. Per-criterion scoring
        contributions: list[CriterionContribution] = []
        for crit in self.criteria:
            contrib = self._score_criterion(crit, project)
            contributions.append(contrib)
            audit.append(
                f"  {contrib.criterion_id}: raw={contrib.raw_score} "
                f"weighted={contrib.weighted_score} ({contrib.explanation})"
            )

        # 3. Combine into base_score
        if self.weighting_model == "formula":
            expr = self.config.get("global_formula", "0")
            vars_map = {c.criterion_id: c.raw_score for c in contributions}
            try:
                base_score = safe_eval(expr, vars_map)
                audit.append(f"global formula '{expr}' = {base_score}")
            except Exception as e:
                base_score = 0
                audit.append(f"global formula error: {e}")
        else:
            base_score = sum(c.weighted_score for c in contributions)

        # 4. Apply adjustments
        final_score, adj_applied, adj_notes = self._apply_adjustments(base_score, project)
        audit.extend(adj_notes)

        # 5. Set-aside
        sa = self._assign_set_aside(project) if passed else None
        if sa:
            audit.append(f"set-aside assignment: {sa}")

        return ProjectScore(
            project_id=project_id,
            project_label=str(project_label),
            eligibility_passed=passed,
            eligibility_failures=failures,
            base_score=base_score,
            contributions=contributions,
            adjustments_applied=adj_applied,
            set_aside_assignment=sa,
            final_score=final_score if passed else 0,
            audit_trail=audit,
        )

    # ── Score a batch / produce ranking ──────────────────────────────
    def rank(self, projects: list[Project]) -> RankedResult:
        scored: list[ProjectScore] = []
        excluded: list[ProjectScore] = []
        for p in projects:
            s = self.score_project(p)
            (excluded if not s.eligibility_passed else scored).append(s)
        return RankedResult(
            model_name=self.metadata["name"],
            model_state=self.metadata["state"],
            weighting_model=self.weighting_model,
            scored=scored,
            excluded=excluded,
        )


# ═══════════════════════════════════════════════════════════════════
#  CSV / JSON LOADERS
# ═══════════════════════════════════════════════════════════════════

def load_projects(path: Union[str, Path]) -> list[Project]:
    """Load projects from CSV or JSON."""
    p = Path(path)
    if p.suffix.lower() == ".json":
        data = json.loads(p.read_text())
        return data if isinstance(data, list) else data.get("projects", [])
    rows: list[Project] = []
    with p.open(newline="", encoding="utf-8") as fh:
        for r in csv.DictReader(fh):
            rows.append({k: _coerce(v) for k, v in r.items()})
    return rows


# ═══════════════════════════════════════════════════════════════════
#  CONVENIENCE
# ═══════════════════════════════════════════════════════════════════

def rank_projects(config_path: Union[str, Path], projects_path: Union[str, Path]) -> RankedResult:
    """One-call helper: load model + projects, return ranked result."""
    model = PriorityModel.from_file(config_path)
    projects = load_projects(projects_path)
    return model.rank(projects)


__all__ = [
    "PriorityModel",
    "ProjectScore",
    "RankedResult",
    "CriterionContribution",
    "rank_projects",
    "load_projects",
    "safe_eval",
]
